iree/runtime/session.h

```diff
@@ -0,0 +1,224 @@
+
+// Returns the HAL device being used for execution.
+//
+// NOTE: this device will not be available until initialized by a user module
+// and will return NULL if queried prior.
+IREE_API_EXPORT iree_hal_device_t* iree_runtime_session_device(
+    const iree_runtime_session_t* session);
+
+// Returns the device allocator used to allocate compatible buffers.
+// Buffers from other allocators may not be compatible and require importing
+// prior to being usable by the session.
+//
+// NOTE: this device allocator will not be available until initialized by a
+// user module and will return NULL if queried prior.
+IREE_API_EXPORT iree_hal_allocator_t* iree_runtime_session_device_allocator(
+    const iree_runtime_session_t* session);
+
+// Appends the given |module| to the context.
+// The module will be retained by the context.
+//
+// NOTE: only valid if the context is not yet frozen; see
+// iree_runtime_session_create for more information.
+IREE_API_EXPORT iree_status_t iree_runtime_session_append_module(
+    iree_runtime_session_t* session, iree_vm_module_t* module);
+
+// Appends a bytecode module to the context loaded from the given memory blob.
+// If a |flatbuffer_allocator| is provided then it will be used to free the
+// |flatbuffer_data| when the module is destroyed and otherwise the ownership of
+// the |flatbuffer_data| remains with the caller. The data must remain valid for
+// the lifetime of the session.
+//
+// If the module exists as a file prefer instead to use
+// iree_runtime_session_append_bytecode_module_from_file to use memory mapped
+// I/O and reduce total memory consumption.
+//
+// NOTE: only valid if the context is not yet frozen; see
+// iree_runtime_session_create for more information.
+IREE_API_EXPORT iree_status_t
+iree_runtime_session_append_bytecode_module_from_memory(
+    iree_runtime_session_t* session, iree_const_byte_span_t flatbuffer_data,
+    iree_allocator_t flatbuffer_allocator);
+
+// Appends a bytecode module to the context loaded from the given |file_path|.
+//
+// NOTE: only valid if the context is not yet frozen; see
+// iree_runtime_session_create for more information.
+IREE_API_EXPORT iree_status_t
+iree_runtime_session_append_bytecode_module_from_file(
+    iree_runtime_session_t* session, const char* file_path);
+
+// Sets |out_function| to to an exported function with the fully-qualified name
+// of |full_name| or returns IREE_STATUS_NOT_FOUND. The function reference is
+// valid for the lifetime of |session|.
+//
+// The function name matches the original MLIR module and function symbols.
+// Example:
+//   module @foo {
+//     func @bar()
+//   }
+// The full name of '@bar' is 'foo.bar'.
+// By default modules have the name 'module'.
+IREE_API_EXPORT iree_status_t iree_runtime_session_lookup_function(
+    const iree_runtime_session_t* session, iree_string_view_t full_name,
+    iree_vm_function_t* out_function);
+
+// Synchronously issues a generic function call.
```

Comment:
Do you see asynchronous calls being supported by iree/runtime/? Or would that be an exclusively low level API?

Reply:
Yep! Probably via `iree_runtime_call_t` (or an async_ variant of it) - that way the user would have a handle to state they could use to query/cancel/etc. I wanted to avoid that here until the ABI refresh is done and we can plumb through semaphores again.